### PR TITLE
[SPARK-39476][SQL] Disable Unwrap cast optimize when casting from Long to Float/ Double or from Integer to Float

### PR DESCRIPTION
**Pull request by WangGuangxin**
### What changes were proposed in this pull request?
Cast from Integer to Float or from Long to Double/Float may loss precision if the length of Integer/Long beyonds the **significant digits** of a Double(which is 15 or 16 digits) or Float(which is 7 or 8 digits).

For example, ```select *, cast(a as int) from (select cast(33554435 as foat) a )``` gives `33554436` instead of `33554435`.

When it comes the optimization rule `UnwrapCastInBinaryComparison`, it may result in incorrect (confused) result .
We can reproduce it with following script.
```
spark.range(10).map(i => 64707595868612313L).createOrReplaceTempView("tbl")
val df = sql("select * from tbl where cast(value as double) = cast('64707595868612313' as double)")
df.explain(true)
df.show()
```

With we disable this optimization rule , it returns 10 records.
But if we enable this optimization rule, it returns empty, since the sql is optimized to
```
select * from tbl where value = 64707595868612312L
``` 

### Why are the changes needed?
Fix the behavior that may confuse users (or maybe a bug?)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a new UT

### CODE REVIEW

**Comment by ulysses-you** [comment]
you should file a new ticket since SPARK-24994 is already released ..

**Comment by WangGuangxin** [comment]
> you should file a new ticket since [SPARK-24994](https://issues.apache.org/jira/browse/SPARK-24994) is already released ..

Thanks for remind, updated. @ulysses-you  @sunchao @dongjoon-hyun @cloud-fan Please help review this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala [hunk 1]
```diff
@@ -358,8 +358,16 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
...
       !fromExp.foldable &&
       toType.isInstanceOf[NumericType] &&
-      ((fromExp.dataType.isInstanceOf[NumericType] && Cast.canUpCast(fromExp.dataType, toType)) ||
-        fromExp.dataType.isInstanceOf[BooleanType])
+      canUnwrapCast(fromExp.dataType, toType)
+  }
+
+  private def canUnwrapCast(from: DataType, to: DataType): Boolean = (from, to) match {
+    case (BooleanType, _) => true
+    case (IntegerType, FloatType) => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala:366]: can we add some code comments to explain the reason?

**Review by cloud-fan — APPROVED** [review]
good catch!

**Review by sunchao — APPROVED** [review]
Good catch! thanks @WangGuangxin !

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3/3.2/3.1!

**Pull request closed without merging**